src/rdkafka_metadata.c

```diff
@@ -1971,3 +1971,110 @@ rd_kafka_metadata_new_topic_with_partition_replicas_mock(int replication_factor,
+
+                for (j = 0; j < mdi->metadata.topics[i].partition_cnt; j++) {
+                        mdp     = &mdi->metadata.topics[i].partitions[j];
+                        mdp_int = &mdi->topics[i].partitions[j];
+
+                        if (mdp->id >= rkmce->rkmce_mtopic.partition_cnt) {
+                                rd_kafka_log(
+                                    rk, LOG_WARNING, "METADATA",
+                                    "Metadata update: Topic %s partition %d "
+                                    "not found in cache",
+                                    rkmce->rkmce_mtopic.topic, mdp->id);
+                                continue;
+                        }
+
+                        rkb = rd_kafka_broker_find_by_nodeid(rk, mdp->leader);
+                        if (!rkb) {
+                                rd_kafka_log(
+                                    rk, LOG_WARNING, "METADATA",
+                                    "Metadata update: Topic %s partition %d "
+                                    "leader %" PRId32 " not found in cache",
+                                    rkmce->rkmce_mtopic.topic, mdp->id,
+                                    mdp->leader);
+                                continue;
+                        }
+
+                        if (rkmce->rkmce_metadata_internal_topic
+                                .partitions[mdp->id]
+                                .leader_epoch >= mdp_int->leader_epoch) {
+                                rd_kafka_dbg(
+                                    rk, METADATA, "METADATA",
+                                    "Cache entry already has newer leader "
+                                    "epoch for topic %s partition %d: ",
+                                    rkmce->rkmce_mtopic.topic, mdp->id);
+                                continue;
+                        }
+                        cache_updated = rd_true;
+
+                        rkmce->rkmce_metadata_internal_topic.partitions[mdp->id]
+                            .leader_epoch = mdp_int->leader_epoch;
+                        rkmce->rkmce_mtopic.partitions[mdp->id].leader =
+                            mdp->leader;
+
+                        rd_kafka_dbg(rk, METADATA, "METADATA",
+                                     "Metadata updating: Topic %s partition %d "
+                                     "with leader %" PRId32
+                                     " and epoch %" PRId32,
+                                     rkmce->rkmce_mtopic.topic, mdp->id,
+                                     mdp->leader, mdp_int->leader_epoch);
+
+                        rd_kafka_metadata_cache_topic_update(
+                            rk, &rkmce->rkmce_mtopic,
+                            &rkmce->rkmce_metadata_internal_topic, rd_true,
+                            rd_true, mdi->brokers, mdi->metadata.broker_cnt,
+                            rd_true);
+                        rd_kafka_topic_metadata_update2(
```

Comment:
This needs to be done once per topic, so you can put it outside the for loop you can have a variable
`partition_cache_changes` that is incremented for each partition and use it in an if outside the partition loop

Reply:
There the rkb is necessary just for logging purposes, you can use `rd_kafka_broker_t *rkb = rk->rk_internal_rkb;` in the if after the partition loop

Reply:
Added a counter `partition_cache_changes` which is incremented for each partition cache change, and if the counter> 0 then the `rd_kafka_topic_metadata_update2` is called.